src/main/java/codeu/controller/ChatServlet.java

```diff
@@ -76,6 +77,33 @@ void setUserStore(UserStore userStore) {
     this.userStore = userStore;
   }
 
+  public static String replaceUrls(String text){
+    String newText = text;
+    ArrayList<String> containedUrls = new ArrayList<String>();
+    Pattern pattern = Pattern.compile(
```

Comment:
I am not a regex expert, can you maybe leave a comment explaining it?

Or consider something simpler, maybe using String's `startWith()` function or something?

Reply:
Or maybe the `replaceAll()` function?

(Note that I'm not saying your way is wrong at all, just wondering about other approaches.)

Reply:
I agree a comment explaining it would be nice here

Reply:
Hey all! I'm actually pretty bad with RegEx but I thought it would be really useful in this case since urls can be typed in so many different ways. RegEx is short for Regular Expressions. They are a sequence of characters that represent a search pattern for a string. I'm using this pattern to find all the locations of urls in the string so that I can replace them. The regex I use above is a really well rated one I found online for urls, because it will match with quite a few url types. Here are a few helpful links I found for reading up on regex or on practicing if you want to use them in other projects:
- [Wiki Article](https://en.wikipedia.org/wiki/Regular_expression)
- [Cheatsheet & Tutorial](https://regexr.com/)

Reply:
I agree that Regex is useful, maybe just add a comment saying this mach a certain type of url, and the links for your sources